src/theory/strings/array_core_solver.h

```diff
@@ -0,0 +1,106 @@
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * Sequences solver for seq.nth/seq.update.
+ */
+
+#include "cvc5_private.h"
+
+#ifndef CVC5__THEORY__STRINGS__ARRAY_CORE_SOLVER_H
+#define CVC5__THEORY__STRINGS__ARRAY_CORE_SOLVER_H
+
+#include "theory/strings/core_solver.h"
+#include "theory/strings/extf_solver.h"
+#include "theory/strings/inference_manager.h"
+#include "theory/strings/solver_state.h"
+#include "theory/strings/term_registry.h"
+
+namespace cvc5 {
+namespace theory {
+namespace strings {
+
+class ArrayCoreSolver : protected EnvObj
+{
+ public:
+  ArrayCoreSolver(Env& env,
+                  SolverState& s,
+                  InferenceManager& im,
+                  TermRegistry& tr,
+                  CoreSolver& cs,
+                  ExtfSolver& es,
+                  ExtTheory& extt);
+  ~ArrayCoreSolver();
+
+  /**
+   * Perform reasoning about seq.nth and seq.update operations.
+   *
+   * Can assume that seq.update / seq.nth terms only apply to concatenation-free
+   * equivalence classes.
+   */
+  void check(const std::vector<Node>& nthTerms,
+             const std::vector<Node>& updateTerms);
+
+  /**
+   *
+   * @param eqc The sequence equivalence class representative. We can assume
+   * the equivalence class of eqc contains no concatenation terms.
+   * @return the map corresponding to the model for eqc. The domain of
+   * the returned map should be in distinct integer equivalence classes of the
+   * equality engine of strings theory. The model assigned to eqc will be
+   * a skeleton constructed via seq.++ where the components take values from
+   * this map.
+   */
+  const std::map<Node, Node>& getWriteModel(Node eqc);
+
+  /**
+   * Get connected sequences
+   * @return a map M such that sequence equivalence class representatives x and
+   * y are connected if an only if M[x] = M[y].
+   */
+  const std::map<Node, Node>& getConnectedSequences();
+
+ private:
+  /** The solver state object */
+  SolverState& d_state;
+  /** The (custom) output channel of the theory of strings */
+  InferenceManager& d_im;
+  /** Reference to the term registry of theory of strings */
+  TermRegistry& d_termReg;
+  /** reference to the core solver, used for certain queries */
+  CoreSolver& d_csolver;
+  /** reference to the extended solver, used for certain queries */
+  ExtfSolver& d_esolver;
+  /** the extended theory object for the theory of strings */
+  ExtTheory& d_extt;
+  /** The write model */
+  std::map<Node, std::map<Node, Node>> d_writeModel;
+  /** Connected */
+  std::map<Node, Node> d_connectedSeq;
+  context::CDHashSet<Node> d_lem;
+
+  // ========= data structure =========
+  std::map<Node, std::set<Node>> d_index_map;
+
+  void sendInference(const std::vector<Node>& exp, const Node& lem);
+
+  void checkNth(const std::vector<Node>& nthTerms);
```

Comment:
Document this method.  It would be helpful if the documentation of the inference was given in the h file instead of the cpp.